src/dpdk_statefulset_updater.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module used to update Kubernetes Statefulset to support DPDK."""
+
+import logging
+from typing import Iterable, Optional
+
+from lightkube import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.models.core_v1 import Container
+from lightkube.resources.apps_v1 import StatefulSet
+
+logger = logging.getLogger(__name__)
+
+
+class DPDKStatefulSetUpdaterError(Exception):
+    """DPDKStatefulSetUpdaterError."""
+
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(self.message)
+
+
+class DPDKStatefulSetUpdater:
+    """Class used to update Kubernetes Statefulset to support DPDK."""
+
+    def __init__(
+        self,
+        statefulset_name: str,
+        namespace: str,
+        dpdk_access_interface_resource_name: str,
+        dpdk_core_interface_resource_name: str,
+    ):
+        self.k8s_client = Client()
+        self.statefulset_name = statefulset_name
+        self.namespace = namespace
+        self.dpdk_resource_requirements = {
+            "requests": {
+                dpdk_access_interface_resource_name: "1",
+                dpdk_core_interface_resource_name: "1",
+            },
+            "limits": {
+                dpdk_access_interface_resource_name: "1",
+                dpdk_core_interface_resource_name: "1",
+            },
+        }
```

Comment:
I am not sure I understand those requirements, what do they represent exactly?

Reply:
Network resources (VFs) are advertised to Kubernetes in the form of allocatable resources. We use [SR-IOV Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) to map the resource to physical PCI address. For configured cluster it should look similar to this:
```
$ kubectl get node  -o json | jq '.items[].status.allocatable'
{
  "cpu": "7",
  "ephemeral-storage": "46779129369",
  "hugepages-1Gi": "2Gi",
  "hugepages-2Mi": "0",
  "intel.com/intel_sriov_vfio_access": "1",
  "intel.com/intel_sriov_vfio_core": "1",
  "memory": "13198484Ki",
  "pods": "110"
}
```
Container's ResourceRequirements are the way to reflect the fact that we need these devices to be available.

Reply:
Ok, so "1" is treated as a boolean in that case by the plugin. Thanks for clarifying!

Reply:
Actually "1" represents quantity of resource of given type. In this case it means we have 1 interface of each type. Number of resources will depend on the number of PCI addresses assigned to a resource.